### DISCUSSION TITLE
How to implement conditional formatting (or templating) for non-YAML content

### DISCUSSION THREAD

**Discussion by lbroudoux** [operator-framework/java-operator-sdk#1840] [Q&A]
Hi,

In my operator reconciliation loop, I need to manage `ConfigMap` that will hold different types of files - typically `.properties` and `.json` content. This content must be customized depending on my Custom Resource: some properties should be set to a CR attribute value, some properties should be added/removed depending on the CR content.

There are straightforward ways to do this in Helm Chart using Go templates and in Ansible Operator SDK using Jinja templates. Is there a recommended/preferred way of doing this in the Java Operator SDK?

I was thinking of loading some template resource as a stream within the reconciler and then using something like [Qute](https://quarkus.io/guides/qute) or [Thymeleaf](https://www.thymeleaf.org/) to produce the final result but I don't know how tied to the HTTP/REST world they are....

Any recommendations or suggestions? Thanks!

**Comment by metacosm** [comment]
You could create a YAML file with that content, load it in your operator and then create a  builder from it and just edit the values that need replacing.

**Reply by lbroudoux** [reply]
Yes that's what I'm doing for regular Kubernetes resources. However it's not convenient to dynamically change properties, json or XML files that are for example wrapped into a `ConfigMap` at hand... there's no builder for that. That's why I'm looking for a more general purpose templating system to build the content to wrap.

**Reply by metacosm** [reply]
How would you like it to look like? Maybe that's something that's general-purpose enough that we could implement it at the JOSDK level?

**Reply by lbroudoux** [reply]
Hi there,

I did a first implementation using [Qute templating](https://quarkus.io/guides/qute) from Quarkus and it's very convenient to use.

From my Java code (in a dependent resource), I just use the `@CheckedTemplate` annotation to get access to a templated JSON file. See [here for an example](https://github.com/microcks/microcks-operator/blob/main/operator/src/main/java/io/github/microcks/operator/resources/KeycloakConfigMapDependentResource.java#L81). 

In this use-case, the `microcksRealm` template is a Keycloak realm definition in JSON format containing templated values based on `spec`. See [here for an example](https://github.com/microcks/microcks-operator/blob/main/operator/src/main/resources/templates/KeycloakConfigMapDependentResource/microcksRealm.json#L183).

I think this could be super-useful to have this at the JOSDK level, may be in the `ReconcilerUtils` class just next to `loadYaml()` method... Something like `ReconcilerUtils.renderTemplate(Class loader, String templateFile, Object spec)` would be nice!

`qute-core` is the only necessary dependency and it looks lightweight. Don't know though if it would be the best candidate...

**Reply by metacosm** [reply]
I will take a look, I have been thinking about this issue on and off for a while actually but I've never really devoted my full attention to it yet. The idea that's been running in my head would be to try to create (at least for the Quarkus side of things) "compiled" templates at build time which would make it efficient to perform these kinds of operations (but also resource cloning) at runtime. It hasn't been a priority so far and I'm actually unsure if there's anything more to it than just a crazy idea 😅

**Comment by csviri** [comment]
Hmm, good question, I usually use mustache if need some simple tamplating:
https://github.com/spullara/mustache.java
